### PR TITLE
Fix bug in submission btc info

### PR DESCRIPTION
**Pull request by KonradStaniec**
While looking for ways of optimizing `BtcCheckpointInfo` endpoint I encountered bug in getting submission btc info, which could lead to misindexing of transaction.

This pr fixes the bug and provides set of tests triggering given problem.

In general I think we need ADR for describing how we decide which submission is the best. It can be done when we will start working on rewards.

### CODE REVIEW

File: x/btccheckpoint/keeper/keeper.go [hunk 1]
```diff
@@ -142,34 +142,57 @@ func (k Keeper) checkSubmissionStatus(ctx sdk.Context, info *types.SubmissionBtc
...
+			// Checkpoint submission `x` for epoch 5, both tx in same block at height 100, with indexes 1 and 10
+			// and
+			// Checkpoint submission `y` for epoch 5, both tx in same block at height 100, with indexes 3 and 9
+			// we will chose submission `x` as the better one.
+			// This good enough solution, but it is not perfect and leads to some edge cases like:
+			// Checkpoint submission `x` for epoch 5, one tx in block 99 with index 1, and second tx in block 100 with index 4
+			// and
+			// Checkpoint submission `y` for epoch 5, both tx in same block at height 100, with indexes 3 and 9
+			// In this case submission `y` will be better as it `earliest` tx in most fresh block is first. But at first glance
+			// submission `x` seems better.
```
Comment (vitsalis) [inline x/btccheckpoint/keeper/keeper.go:182]: How about we collect sets of `(firstPartBlockHeight, secondPartBlockHeight, firstPartTxIndex, secondPartTxIndex)` and then sort them based on an algorithm we define to choose the values that refer to the best submission? Will allow us to cleanly separate the selection algorithm and make it reusable by other people that are looking into the BTC chain to select the best Babylon chain. The second is especially very powerful as we want people to be able to efficiently differentiate the proper Babylon chain.

I think that in this case (even though those are the same checkpoints), we should clearly choose `y`. However, as long as everyone uses the same selection rule as us, we're good.

Comment (KonradStaniec) [reply x/btccheckpoint/keeper/keeper.go:182]: So there is already struct which is used for that purpose with `IsBetterThan` defined - https://github.com/babylonchain/babylon/blob/dev/x/btccheckpoint/types/types.go#L138. It does not contain all the data you mentioned as not all data is used currently for scoring purposes.

Current algorithm is pretty simple but definitely not the best i.e submission `A` is considered  better than submission `B` , if its most recent header is deeper in btc chain that submission `B` most recent header. In case of a tie (most recent header is the same one), submission with lowest txidx in most recent header wins.

In general I can find both advantages and disadvantages of this approach. (and several possible alternatives)

Imo current route for this problem is:
1. merge this pr as it fixes pretty insidious bug
2. create ADR/RFC spelling out possible alternatives. Also it would be nice to connect this problem to rewards because as soon you introduce rewards and relaying on transaction ordering to get those rewards, you must consider MEV i.e miners observing transactions in mempool and reordering them/injecting their own transactions to get those rewards.
3. After this ADR we can intrduce necessary data to `SubmissionBtcInfo` and tweak `IsBetterThan` implementation.

Comment (vitsalis) [reply x/btccheckpoint/keeper/keeper.go:182]: Makes sense, thanks for the explanation!

**Review by vitsalis — APPROVED** [review]

File: x/btccheckpoint/keeper/keeper_test.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		expectedOldestBlockDepth   uint64
+	}{
+		{"First header older. TxIndex larger in older header.", args{TxKeyDesc{TxIdx: 5, Depth: 10}, TxKeyDesc{TxIdx: 1, Depth: 0}}, 0, 1, 10},
+		{"First header older. TxIndex larger in younger header.", args{TxKeyDesc{TxIdx: 1, Depth: 10}, TxKeyDesc{TxIdx: 5, Depth: 0}}, 0, 5, 10},
+		{"Second header older. TxIndex larger in older header.", args{TxKeyDesc{TxIdx: 1, Depth: 0}, TxKeyDesc{TxIdx: 5, Depth: 10}}, 0, 1, 10},
+		{"Second header older. TxIndex larger in younger header.", args{TxKeyDesc{TxIdx: 5, Depth: 0}, TxKeyDesc{TxIdx: 1, Depth: 10}}, 0, 5, 10},
+		{"Same block. TxIndex larger in first transaction key.", args{TxKeyDesc{TxIdx: 5, Depth: 10}, TxKeyDesc{TxIdx: 1, Depth: 10}}, 10, 1, 10},
+		{"Same block. TxIndex larger in second transaction key.", args{TxKeyDesc{TxIdx: 1, Depth: 10}, TxKeyDesc{TxIdx: 5, Depth: 10}}, 10, 1, 10},
+	}
+	for _, tt := range tests {
```
Comment (vitsalis) [inline x/btccheckpoint/keeper/keeper_test.go:38]: Can we make this a fuzz test? Given that it depends on randomness.

Comment (KonradStaniec) [reply x/btccheckpoint/keeper/keeper_test.go:38]: tbh the only randomness here is for hashes for which we do not care that much. And I would prefer to have this particular test cases always checked as they trigger the bug. 

On the other hand fuzz test here is pretty good idea to check different combinations.

tldr; I have added separate fuzzing test for this method 😅

File: x/btccheckpoint/keeper/keeper_test.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+			require.Equal(t, info.LatestTxIndex, tt.expectedTxIndex, tt.name)
+			require.Equal(t, info.OldestBlockDepth, tt.expectedOldestBlockDepth, tt.name)
+		})
+	}
+}
+
+func FuzzGetSubmissionBtcInfo(f *testing.F) {
+	f.Add(int64(1), uint32(0), uint32(1), uint32(1), uint32(1))
+
+	f.Fuzz(func(t *testing.T, seed int64, depth1 uint32, txidx1 uint32, depth2 uint32, txidx2 uint32) {
```
Comment (vitsalis) [inline x/btccheckpoint/keeper/keeper_test.go:69]: The style that we typically follow is we generate all the parameters from the randomness seed. e.g. we could have the following

```
depth1 := r.Uint32()
depth2 := r.Uint32()
txidx1 := r.Uint32()
txidx2 := r.Uint32()
for txidx2 == txidx1 {
    txidx2 = r.Uint32()
}
```

File: x/btccheckpoint/keeper/keeper_test.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+			require.Equal(t, info.YoungestBlockDepth, tt.expectedYoungestBlockDepth, tt.name)
+			require.Equal(t, info.LatestTxIndex, tt.expectedTxIndex, tt.name)
+			require.Equal(t, info.OldestBlockDepth, tt.expectedOldestBlockDepth, tt.name)
+		})
+	}
+}
+
+func FuzzGetSubmissionBtcInfo(f *testing.F) {
+	f.Add(int64(1), uint32(0), uint32(1), uint32(1), uint32(1))
```
Comment (vitsalis) [inline x/btccheckpoint/keeper/keeper_test.go:67]: If we had only the `seed` parameter here, we could dynamically generate `n` test cases each time with `datagen.AddRandomSeedsToFuzzer(f, n)`

**Review by vitsalis — COMMENTED** [review]
Fuzz test looks good, although it does not follow our usual style. I guess this is related to our discussion about letting the Go fuzzer try to cover all paths.

File: x/btccheckpoint/keeper/keeper_test.go (hunk 3 above)
Comment (KonradStaniec) [reply x/btccheckpoint/keeper/keeper_test.go:69]: hmm I think I missed this parts of discussion. How the style you describe is better than just go fuzzer generating this values for us ?

Comment (vitsalis) [reply x/btccheckpoint/keeper/keeper_test.go:69]: The reasoning is that we generate ourselves all of the data that we need for the tests through datagen methods and therefore we only need a seed. This allows us to only generate valid data (for example in this case different tx idxs) without having to skip due to the Go fuzzer providing invalid data. In this particular case this usefulness is not that obvious as we will rarely have to skip (given to how easy it is to generate different tx idxs), but for more complex data types (e.g. a BTC hash) we would skip most of the tests if we didn't generate proper data ourselves.

Comment (vitsalis) [reply x/btccheckpoint/keeper/keeper_test.go:69]: No problem with the way the test is structured right now, but I would prefer if it followed the style we have in the rest of the codebase.

Comment (KonradStaniec) [reply x/btccheckpoint/keeper/keeper_test.go:69]: Seems reasonable. I will adapt the test so it will not stand out  👍
